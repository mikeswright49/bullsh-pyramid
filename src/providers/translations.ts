export const TRANSLATIONS = {
    en: {
        'game.board.current.row': 'Current Row: {{ activeRow }}',
        'game.board.current.card': 'Current card',
        'game.display.stage.memorization.text': 'You have 30s to remember your cards!',
        'game.display.stage.flipping.text': 'The host is flipping a card',
        'game.display.stage.declaration.text': 'Let the world know if you have it',
        'game.display.stage.assign.text': 'Players are beginning to lie left right and center',
        'game.display.stage.bullshit.text': 'People should know who you think was lieing',
        'game.display.stage.reveal.text': 'We all now know the truth',
        'game.display.stage.memory.text': 'Well which cards did you have?',
        'game.display.stage.complete.text': 'Well time for a new game',
    },
};
