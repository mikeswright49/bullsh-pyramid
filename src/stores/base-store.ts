import firebase from 'firebase';
import { getFirebaseConfig } from 'config/firebase-config';

export abstract class BaseStore {
    public static database: firebase.database.Database;

    public static init() {
        if (!firebase.apps.length) {
            firebase.initializeApp(getFirebaseConfig());
            firebase.auth().signInAnonymously();
        }
        if (!BaseStore.database) {
            BaseStore.database = firebase.database();
        }
    }
}
