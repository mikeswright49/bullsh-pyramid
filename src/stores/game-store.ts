import * as firebase from 'firebase';
import { shortId } from '../utilities/shortid';
import { GameStage } from '../enums/game-stage';
import { GameState } from 'types/game-state';
import { getFirebaseConfig } from '../../config/firebase-config';
import { Player } from 'types/player';
import { Card } from 'types/card';

export class GameStore {
    private static database: firebase.database.Database;
    private static subscribers = [];

    public static async createGame(playerCount: number, tierCount: number): Promise<string> {
        GameStore.init();
        const ID_LENGTH = 4;
        const gameId = shortId(ID_LENGTH);

        const gameState = {
            id: gameId,
            gameStage: GameStage.Initiation,
            activeIndex: 0,
            activeRow: 0,
            playerCount,
            tierCount,
        } as GameState;

        try {
            await GameStore.database.ref(`games/${gameId}`).set(gameState);
            return gameId;
        } catch (error) {
            console.error(error);
        }
    }

    public static subscribeToGame(gameId: string, updateCallback: (snapshot: GameState) => void) {
        GameStore.init();
        const game = GameStore.database.ref(`games/${gameId}`);
        GameStore.subscribers.push(
            game.on('value', (val) => {
                const value = val.val();
                updateCallback(value as GameState);
            })
        );
    }

    public static subscribeToPlayers(gameId: string, updateCallback: (snapshot: Player) => void) {
        GameStore.init();
        const game = GameStore.database.ref(`games/${gameId}/players`);
        const players = GameStore.database.ref(`players`);
        game.on('child_added', (child) => {
            const playerId = child.val();
            players.child(playerId).on('value', (val) => {
                const value = val.val();
                updateCallback(value as Player);
            });
        });
    }

    public static unsubscribeToGame() {
        GameStore.subscribers.forEach((sub) => sub.off('value'));
    }

    public static async updateGameStage(gameId: string, gameStage: GameStage) {
        await GameStore.database.ref(`games/${gameId}`).update({
            gameStage,
        });
    }

    public static async updateTiers(gameId: string, tiers: Card[][]) {
        await GameStore.database.ref(`games/${gameId}`).update({
            tiers,
        });
    }

    public static async setActiveCard(
        gameId: string,
        index: { activeRow: number; activeIndex: number }
    ) {
        await GameStore.database.ref(`games/${gameId}`).update(index);
    }

    public static async joinGame(gameId: string, playerId: string) {
        await GameStore.database.ref(`games/${gameId}/players`).push(playerId);
    }

    private static init() {
        if (!firebase.apps.length) {
            firebase.initializeApp(getFirebaseConfig());
        }
        if (!GameStore.database) {
            GameStore.database = firebase.database();
        }
    }
}
