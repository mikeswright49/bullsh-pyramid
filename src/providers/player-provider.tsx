import React from 'react';
import { useState, useEffect } from 'react';
import { Player } from 'types/player';
import { PlayerStore } from 'src/stores/player-store';
export const PlayerContext = React.createContext<Player>(null);

export function PlayerProvider(props: React.PropsWithChildren<{ playerId: string }>) {
    const [player, setPlayer] = useState<Player>();

    useEffect(() => {
        if (!props.playerId) {
            return;
        }
        PlayerStore.subscribeToPlayer(props.playerId, (updatedPlayer) => {
            setPlayer(updatedPlayer);
        });

        return () => PlayerStore.unsubscribeToPlayer();
    }, [props.playerId]);

    return <PlayerContext.Provider value={player}>{props.children}</PlayerContext.Provider>;
}
