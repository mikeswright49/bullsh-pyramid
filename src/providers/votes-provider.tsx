import React from 'react';
import { useState, useEffect } from 'react';
import { Vote } from 'types/vote';
import { VoteStore } from 'src/stores/vote-store';
export const VotesContext = React.createContext<Vote[]>([]);

export function VotesProvider(props: React.PropsWithChildren<{ gameId: string }>) {
    const [votes, setVotes] = useState([]);
    useEffect(() => {
        if (!votes) {
            return;
        }

        VoteStore.subscribeToVotes(props.gameId, (vote: Vote) => {
            votes[vote.id] = vote;
            setVotes({ ...votes });
        });
    }, [props.gameId]);

    return (
        <VotesContext.Provider value={Object.values(votes)}>{props.children}</VotesContext.Provider>
    );
}
