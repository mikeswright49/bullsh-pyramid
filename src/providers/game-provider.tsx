import React, { useEffect, useState } from 'react';
import { GameState } from 'types/game-state';
import { GameStore } from 'src/stores/game-store';
import { GameStage } from 'src/enums/game-stage';

export const DEFAULT_GAME_STATE: GameState = {
    id: '',
    gameStage: GameStage.Initiation,
    activeRow: 0,
    activeIndex: 0,
    cardCount: 4,
    flipDelay: 5000,
    players: [],
    tiers: [],
};
export const GameContext = React.createContext<GameState>(DEFAULT_GAME_STATE);

export function GameProvider(props: React.PropsWithChildren<{ gameId: string }>) {
    const { gameId } = props;
    const [gameState, setGameState] = useState<GameState>(DEFAULT_GAME_STATE);

    useEffect(() => {
        if (!gameId) {
            return;
        }

        GameStore.subscribeToGame(gameId, (updatedGameState) => {
            setGameState(updatedGameState);
        });

        return () => GameStore.unsubscribeToGame();
    }, [gameId]);

    return <GameContext.Provider value={gameState}>{props.children}</GameContext.Provider>;
}
