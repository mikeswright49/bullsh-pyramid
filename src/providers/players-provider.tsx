import React from 'react';
import { useState, useEffect } from 'react';
import { Player } from 'types/player';
import { GameStore } from 'src/stores/game-store';

export const PlayersContext = React.createContext<Player[]>([]);

export function PlayersProvider(props: React.PropsWithChildren<{ gameId: string }>) {
    const [players, setPlayers] = useState({});
    useEffect(() => {
        GameStore.subscribeToPlayers(props.gameId, (player: Player) => {
            players[player.id] = player;
            setPlayers({ ...players });
        });
    }, [props.gameId]);
    return (
        <PlayersContext.Provider value={Object.values(players)}>
            {props.children}
        </PlayersContext.Provider>
    );
}
