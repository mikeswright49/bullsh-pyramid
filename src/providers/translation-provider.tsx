import React, { useContext } from 'react';
import { useEffect, useState } from 'react';
import { GameContext } from './game-provider';
import { TRANSLATIONS } from '../translations/translations';

const INTERPOLATE_REGEX = /{{.*?}}/gi;
const INTERPOLATE_PROPERTY_REGEX = /{{(.*)?}}/i;
const INTERPOLATE_CAPTURE_COUNT = 2;

export type Translations = {
    translations: { [key: string]: string };
    translate: (key: string, variables?: { [key: string]: string | number | boolean }) => string;
};

export const TranslationContext = React.createContext<Translations>({
    translate: (a) => a,
    translations: {},
});

export function TranslationProvider(props: React.PropsWithChildren<unknown>) {
    const { language } = useContext(GameContext);
    const [translations, setTranslations] = useState({});
    useEffect(() => {
        setTranslations(TRANSLATIONS[language || 'en']);
    }, [language]);

    /**
     * Returns the nested property value (userMetadata.giftcard.bacon)
     * @param source property value
     */
    function getProperty(source: string[], parent: unknown): string {
        if (!source || !parent) {
            return;
        }
        if (source.length > 1) {
            const property: string = source.shift();
            return this.getProperty(source, parent[property.trim()]);
        }
        return parent[source[0].trim()];
    }

    /**
     * Evaluates the {{}} interpolated strings
     * @param source Source content to be parsed
     * @param data Data object to read values off of
     */
    function evaluateInterpolate(
        source: string,
        data: { [key: string]: string | number | boolean }
    ): string {
        if (!data) {
            return source;
        }

        const matches = source.match(INTERPOLATE_REGEX) || [];
        matches.forEach((match: string) => {
            const properties = INTERPOLATE_PROPERTY_REGEX.exec(match);
            if (properties && properties.length === INTERPOLATE_CAPTURE_COUNT) {
                const property = getProperty(properties[1].split('.'), data);
                source = source.replace(match, property || '');
            }
        });
        return source;
    }

    function translate(key: string, variables?: { [key: string]: string | number | boolean }) {
        if (!translations) {
            return '';
        }
        return evaluateInterpolate(translations[key], variables);
    }

    return (
        <TranslationContext.Provider value={{ translations, translate }}>
            {props.children}
        </TranslationContext.Provider>
    );
}
