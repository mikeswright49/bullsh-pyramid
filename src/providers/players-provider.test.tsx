import React from 'react';
import { useContext } from 'react';
import { PlayersContext } from 'src/providers/players-provider';
import { PlayersProvider } from './players-provider';
import { render, act } from '@testing-library/react';
import { GameStore } from 'src/stores/game-store';
import { MOCK_PLAYER } from 'src/__tests__/data/mock-player';

jest.mock('src/stores/game-store');
describe('<Unit Test> Players Provider', () => {
    function ComponentMock() {
        const data = useContext(PlayersContext);
        return <div>{JSON.stringify(data)}</div>;
    }

    it('should subscribe to the players from the store, and map the players to an array it as a value', async () => {
        let cb;
        (GameStore.subscribeToPlayers as jest.Mock) = jest.fn((id, _cb) => (cb = _cb));

        const { container } = render(
            <PlayersProvider gameId="123">
                <ComponentMock />
            </PlayersProvider>
        );

        act(() => cb(MOCK_PLAYER));

        expect(GameStore.subscribeToPlayers).toHaveBeenCalledWith('123', expect.any(Function));
        expect(container).toMatchSnapshot();
    });
});
