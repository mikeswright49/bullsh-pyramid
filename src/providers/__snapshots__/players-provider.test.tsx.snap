// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Unit Test> Players Provider should subscribe to the players from the store, and map the players to an array it as a value 1`] = `
<div>
  <div>
    [{"hand":[{"hidden":true,"suit":2,"value":7},{"hidden":true,"suit":3,"value":8},{"hidden":true,"suit":1,"value":11},{"hidden":true,"suit":2,"value":0}],"id":"nez","isHost":true,"name":"frank","score":10000,"declaration":[],"voteSubmitted":0}]
  </div>
</div>
`;
